banyand/stream/query.go

```diff
@@ -0,0 +1,361 @@
+	if !qr.loaded {
+		if len(qr.data) == 0 {
+			return nil
+		}
+		// TODO:// Parallel load
+		tmpBlock := generateBlock()
+		defer releaseBlock(tmpBlock)
+		for i := 0; i < len(qr.data); i++ {
+			if !qr.data[i].loadData(tmpBlock) {
+				qr.data = append(qr.data[:i], qr.data[i+1:]...)
+				i--
+			}
+			if qr.orderByTimestampDesc() {
+				qr.data[i].idx = len(qr.data[i].timestamps) - 1
+			}
+		}
+		qr.loaded = true
+		heap.Init(qr)
+	}
+	if len(qr.data) == 0 {
+		return nil
+	}
+	if len(qr.data) == 1 {
+		r := &pbv1.Result{}
+		bc := qr.data[0]
+		bc.copyAllTo(r)
+		qr.data = qr.data[:0]
+		return r
+	}
+	return qr.merge()
+}
+
+func (qr *queryResult) Release() {
+	for i, v := range qr.data {
+		releaseBlockCursor(v)
+		qr.data[i] = nil
+	}
+	qr.data = qr.data[:0]
+	for i := range qr.pws {
+		qr.pws[i].decRef()
+	}
+	qr.pws = qr.pws[:0]
+}
+
+func (qr queryResult) Len() int {
+	return len(qr.data)
+}
+
+func (qr queryResult) Less(i, j int) bool {
+	leftTS := qr.data[i].timestamps[qr.data[i].idx]
+	rightTS := qr.data[j].timestamps[qr.data[j].idx]
+	leftVersion := qr.data[i].p.partMetadata.Version
+	rightVersion := qr.data[j].p.partMetadata.Version
+	if qr.orderByTS {
+		if leftTS == rightTS {
+			if qr.data[i].bm.seriesID == qr.data[j].bm.seriesID {
+				// sort version in descending order if timestamps and seriesID are equal
+				return leftVersion > rightVersion
+			}
+			// sort seriesID in ascending order if timestamps are equal
+			return qr.data[i].bm.seriesID < qr.data[j].bm.seriesID
+		}
+		if qr.ascTS {
+			return leftTS < rightTS
+		}
+		return leftTS > rightTS
+	}
+	leftSIDIndex := qr.sidToIndex[qr.data[i].bm.seriesID]
+	rightSIDIndex := qr.sidToIndex[qr.data[j].bm.seriesID]
+	if leftSIDIndex == rightSIDIndex {
+		if leftTS == rightTS {
+			// sort version in descending order if timestamps and seriesID are equal
+			return leftVersion > rightVersion
+		}
+		// sort timestamps in ascending order if seriesID are equal
+		return leftTS < rightTS
+	}
+	return leftSIDIndex < rightSIDIndex
+}
+
+func (qr queryResult) Swap(i, j int) {
+	qr.data[i], qr.data[j] = qr.data[j], qr.data[i]
+}
+
+func (qr *queryResult) Push(x interface{}) {
+	qr.data = append(qr.data, x.(*blockCursor))
+}
+
+func (qr *queryResult) Pop() interface{} {
+	old := qr.data
+	n := len(old)
+	x := old[n-1]
+	qr.data = old[0 : n-1]
+	return x
+}
+
+func (qr *queryResult) orderByTimestampDesc() bool {
+	return qr.orderByTS && !qr.ascTS
+}
+
+func (qr *queryResult) merge() *pbv1.Result {
+	step := 1
+	if qr.orderByTimestampDesc() {
+		step = -1
+	}
+	result := &pbv1.Result{}
+	var lastPartVersion int64
+	var lastSid common.SeriesID
+
+	for qr.Len() > 0 {
+		topBC := qr.data[0]
+		if lastSid != 0 && topBC.bm.seriesID != lastSid {
+			return result
+		}
+		lastSid = topBC.bm.seriesID
+
+		if len(result.Timestamps) > 0 &&
+			topBC.timestamps[topBC.idx] == result.Timestamps[len(result.Timestamps)-1] {
+			if topBC.p.partMetadata.Version > lastPartVersion {
```

Comment:
Please remove the deduplication piece because the stream could store duplicated elements.